### PR TITLE
[AutoDiff] TF-508: Temporary Hack Fix

### PR DESCRIPTION
**Pull request by bartchr808**
[JIRA Ticket TF-508](https://bugs.swift.org/browse/TF-508)

This ticket should not be closed, as this PR is a temporary hack to fix the problem and unblock me on `SIMD` vectors.

The problem has to do with not being able to resolve `Self` for the actual underlying struct type. For example, in the reproducer in the JIRA ticket, it doesn't know about `MyStruct`. So here, we take a look at the first type, which we are assuming is `Self`, but more specfically the struct which lets us successfully check the requirements.

Note that this only fixes methods (notice the `original->hasSelfParam()`), meaning this problem can come back if we were to create a similar reproducer as in the JIRA ticket, but instead on functions where there is no `Self` type.

### CODE REVIEW

**Comment by rxwei** [comment]
Could you add a test case for it?

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -2008,7 +2025,7 @@ static SubstitutionMap getSubstitutionMap(
 /// Returns `None` on failure, signifying that a diagnostic has been emitted.
 ///
 /// Creates new differentiation tasks, if necessary, using `invoker` as the
-/// invoker. Calls `taskCallback` for all newly-created tasks (but may also call
+/// invoker calls `taskCallback` for all newly-created tasks (but may also call
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Revert this change?

File: test/AutoDiff/generics.swift [hunk 2]
```diff
@@ -57,25 +57,76 @@ struct SupervisedTrainer<Model : Layer> {
...
-    func applied(to input: TF_440_Input<T, Float>) -> Float {
-        return input.state
-    }
-
-    @differentiable
-    func applied(to input: TF_440_Input<T, Float>) -> T {
-        return input.input
-    }
+  @differentiable
+  func applied(to input: TF_440_Input<Float, Float>) -> Float {
```
Comment (bartchr808) [inline test/AutoDiff/generics.swift]: Spacing was 4 and not in the 80 column limit, so I thought I'd update it here right away. Let me know if this is fine or if I should pull it out into a separate PR.

File: test/AutoDiff/generics.swift [hunk 3]
```diff
@@ -57,25 +57,75 @@ struct SupervisedTrainer<Model : Layer> {
...
+    return input.state
+  }
+
+  @differentiable
+  func applied(to input: TF_440_Input<T, Float>) -> T {
+    return input.input
+  }
+}
+
+// Tests TF-508 (Hack)
```
Comment (rxwei) [inline test/AutoDiff/generics.swift]: You can drop `(Hack)` from here since the test is not a hack (the implementation is).

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

File: test/AutoDiff/generics.swift (hunk 3 above)
Comment (bartchr808) [reply test/AutoDiff/generics.swift]: Oh ya true! I just wanted to differentiate it for when this bug is fully fixed, but ya the test doesn't matter.

**Review by rxwei — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
The test didn't pass type checking.

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow Linux

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow Linux
